File: executor/src/main/java/com/codeloom/executor/service/EventService.java
package com.codeloom.executor.service;

import com.codeloom.common.SubmissionStatus;
import com.codeloom.common.event.*;
import com.codeloom.executor.engine.SubmissionContext;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.stereotype.Service;
import tools.jackson.databind.ObjectMapper;

@Service
public class EventService {
  private final KafkaTemplate<String, String> kafka;
  private final String topic;
  private final ObjectMapper mapper;

  public EventService(
      KafkaTemplate<String, String> k,
      @Value("${codeloom.kafka.topics.submission-status}") String t,
      ObjectMapper m) {
    kafka = k;
    topic = t;
    mapper = m;
  }

  public void submissionStatusChanged(SubmissionContext c, SubmissionStatus s) {
    submissionStatusChanged(c, s, null);
  }

  public void submissionStatusChanged(
      SubmissionContext c, SubmissionStatus s, SubmissionStatusPayload p) {
    var e = new SubmissionStatusChangedEvent(c.submissionId(), c.userId(), c.problemId(), s, p);
    kafka.send(topic, c.submissionId().toString(), mapper.writeValueAsString(e));
  }
}

File: executor/src/main/kotlin/com/codeloom/executor/service/EventService.kt
package com.codeloom.executor.service

import com.codeloom.common.SubmissionStatus
import com.codeloom.common.event.SubmissionStatusChangedEvent
import com.codeloom.common.event.SubmissionStatusPayload
import com.codeloom.executor.engine.SubmissionContext
import org.springframework.beans.factory.annotation.Value
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.stereotype.Service
import tools.jackson.module.kotlin.jacksonObjectMapper

@Service
class EventService(
    private val kafkaTemplate: KafkaTemplate<String, String>,
    @Value("\${codeloom.kafka.topics.submission-status}")
    private val submissionStatusTopic: String,
) {
    fun submissionStatusChanged(
        context: SubmissionContext,
        newStatus: SubmissionStatus,
        payload: SubmissionStatusPayload? = null,
    ) {
        val event =
            SubmissionStatusChangedEvent(
                submissionId = context.submissionId,
                userId = context.userId,
                problemId = context.problemId,
                newStatus = newStatus,
                payload = payload,
            )

        kafkaTemplate.send(
            submissionStatusTopic,
            context.submissionId.toString(),
            jacksonObjectMapper().writeValueAsString(event),
        )
    }
}
